src/hooks/useForm.js

```diff
@@ -0,0 +1,218 @@
+   * @param {Event} event - The onChange event.
+   * @returns {void} Nothing.
+   */
+  function handleInputChange(event) {
+    const { form } = state;
+    const { value, name } = event.target;
+
+    // Ah, the good old days!
+    setState({
+      ...state,
+      form: {
+        ...form,
+        [name]: value
+      },
+      errors: {
+        ...validateProperty(name, value)
+      }
+    });
+  }
+
+  /**
+   * Validate one or pair of corresponding properties.
+   * @param {string} name - The property name.
+   * @param {string} value - The property value.
+   * @returns {object} All the errors in the entire form.
+   */
+  function validateProperty(name, value) {
+    const { errors, form } = state;
+    let allErrors = {
+      ...errors
+    };
+    const matchingProperty = getMatchingProperty(name, schema);
+
+    // No matching property, just validate this one property
+    if (!matchingProperty) {
+      const { isValid, errors: propertyErrors } = validate(value, schema[name]);
+
+      isValid
+        ? delete allErrors[name]
+        : (allErrors = {
+            ...allErrors,
+            [name]: propertyErrors
+          });
+
+      return { ...allErrors };
+    }
+
+    // Matching properties present. ex: password & confirm password
+    const matchingValues = {
+      [name]: value,
+      [matchingProperty]: form[matchingProperty]
+    };
+
+    const matchingValuesSchema = {
+      [name]: schema[name],
+      [matchingProperty]: schema[matchingProperty]
+    };
+
+    // Clear previous errors on matching properties before
+    // potentially re-setting them
+    delete allErrors[name];
+    delete allErrors[matchingProperty];
+
+    const { errors: propertyErrors } = validate(matchingValues, matchingValuesSchema);
+
+    return {
+      ...allErrors,
+      ...propertyErrors
+    };
+  }
+
+  return {
+    ...state,
+    handleReset,
+    handleSubmit,
+    handleInputChange
+  };
+}
+
+/**
+ * Derive state from the given schema.
+ * @param {object} schema - The given schema.
+ * @param {object} initialFormState - The initial values of the form properties.
+ * @returns {...UseFormState} The useForm initial state.
+ */
+function init(schema, initialFormState) {
+  let form = initialFormState;
+
+  if (!form) {
+    form = {};
+    for (const property in schema) {
+      if (schema.hasOwnProperty(property)) {
+        form[property] = EMPTY_VALUE;
+      }
+    }
+  }
+  return { form, errors: {}, submitError: EMPTY_VALUE };
+}
+
+/**
+ * Get the corresponding property that matches the
+ * current property being validated.
+ *
+ * @param {string} name - The property being validated.
+ * @param {object} schema - The schema of the entire form.
+ * @return {string} The name of the matching property.
+ */
+function getMatchingProperty(name, schema) {
+  const { matchingProperty } = schema[name];
+
+  if (matchingProperty) return matchingProperty;
+
+  for (const property in schema) {
```

Comment:
this for each confuses me, maybe i need to see it in action

Reply:
I'm going to rename `schema` to `formSchema` just to be clearer. It's matching up elements that have the `matches` option. For example: 
```javascript
const formSchemas = 
{
    pwd: new Schema(), 
    confPwd: new Schema().matches('pwd')
}; 
```
When **confPwd** fires an onchange event, validation is going to look at its schema and say ok **confPwd** matches up to **pwd** and it just plucks **pwd**'s schema from the **formSchemas**. But when **pwd** fires an onchange event, its schema does NOT have a `matches` property on it.

The loop is going through all the **formSchemas** checking to see if any singular schema has a `matches` property whose value is **pwd**, and in this particular case **confPwd**'s schema does contain exactly that.

By doing this, we're forcing validation for both elements each time one of them changes.